overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
+func Import(ctx context.Context, id uint64, r io.Reader) (size int64, snapNames []string, err error) {
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", fmt.Sprintf("import-%d", id))
+	if err := os.MkdirAll(path.Join(dirs.SnapCacheDir, "snapshots"), 0755); err != nil {
+		return 0, nil, fmt.Errorf("cannot create snapshots directory")
+	}
+	if err := os.Mkdir(p, 0755); err != nil {
+		if os.IsExist(err) {
+			return 0, nil, fmt.Errorf("snapshot import already in progress for ID `%d`", id)
+		}
+		return 0, nil, fmt.Errorf("failed creating import cache: %v", err)
+	}
+	defer os.RemoveAll(p)
+
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
+	if err != nil {
+		return 0, nil, err
+	}
+	if !exportFound {
+		return 0, nil, fmt.Errorf("snapshot import file incomplete: no export.json file")
+	}
+
+	// walk the cache directory to store the files
+	dir, err := osOpen(p)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed opening import cache for %s: %v", comment, err)
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			names, err := moveCachedSnapshots(names, id, p)
+			if err != nil {
+				return 0, nil, err
+			}
+			// should we simply read all dir names at once above?
+			snapNames = append(snapNames, names...)
+		}
+	}
+	if readErr != nil && readErr != io.EOF {
+		return 0, nil, fmt.Errorf("failed read from import cache for %s: %v", comment, readErr)
+	}
+	return size, snapNames, err
+}
+
+func unpackSnapshotImport(r io.Reader, p string) (exportFound bool, size int64, err error) {
+	tr := tar.NewReader(r)
+	var tarErr error
+	var header *tar.Header
+
+	for tarErr == nil {
+		header, tarErr = tr.Next()
+		if tarErr == io.EOF {
+			break
+		}
+		switch {
+		case tarErr != nil:
+			return false, 0, fmt.Errorf("failed reading snapshot import: %v", tarErr)
+		case header == nil:
+			// should not happen
+			return false, 0, fmt.Errorf("tar header not found")
+		case header.Typeflag == tar.TypeDir:
+			return false, 0, errors.New("unexpected directory in import file")
+		}
+
+		if header.Name == "export.json" {
+			exportFound = true
+		}
+
+		target := path.Join(p, header.Name)
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+		if err != nil {
+			return false, 0, fmt.Errorf("failed creating file `%s`: %v", target, err)
+		}
+		defer t.Close()
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(names []string, id uint64, p string) ([]string, error) {
```

Comment:
This should actually just move things around from one dir to another, currently it's doing much more (validating, renaming).